Refactor to use GUIDs (and fix asset rename/move)

rhys-vdw opened the pull request:
Hey @ogxd. I've switched all storage of paths to GUID and it seems to work well (and fixes #15).

~~**Please don't merge yet!**~~

~~I haven't given it a super thorough test, but I thought I'd share now in case you have any feedback.~~

~~Biggest issue is that the warnings now all log out GUIDs instead of readable file paths.~~

~~This is my plan for warnings~~

TODO:
- [x] `AssetModificationProcessor` should pass paths, not GUIDs, then the warnings will have the correct name for the files changing.
- [x] In the case of dependencies/referencers, if they still exist then we should be able to use `AssetDatabase.GUIDToAssetPath` to render the correct string.
- [x] In the case that we have a GUID that references a non-existent asset, just write "Missing asset (\<GUID>)" or similar. Slightly worse but those warnings are really just there for debugging this tool, not for the end user.

Anyway, let me know what you think.

rhys-vdw committed:
Quick and dirty GUID refactor

rhys-vdw committed:
Fix issue where creating assets would error

rhys-vdw committed:
Show asset paths in warnings

rhys-vdw committed:
Prefix warnings with ProjectCurator

rhys-vdw commented:
Hey @ogxd I've addressed all the issues above and given it a bit more of a test.

I think this all works. But just to be sure I'll set this up as a dependency for our project and run with it for a bit to see if any issues crop up. I'll let you know how it goes over the next week or so. I'll also keep an eye on this in case you have anything you'd like changed.

You're welcome to test and merge if you're happy with it.

ogxd commented:
Thanks @rhys-vdw ! I'll have a look this week and check if I can find any issues with the changes

rhys-vdw commented:
Had no issues with this yet. My only thought is maybe it'd be worth renaming this method:

`FileInfo.GetDependencies()` -> `FileInfo.GetDependencyPaths()`

Since everything else is now GUID-based.

Also this will be a breaking change to anyone using the public API so it would be a major version bump. I'm assuming this is niche enough that I don't need to add back-compatibility...

rhys-vdw commented:
Ah scratch that, it seems project overlay is broken.

rhys-vdw committed:
Fix project overlay

Project overlay correctly uses GUID argument.

rhys-vdw committed:
Remove AssetInfo.GetDependencies

Because this method returns paths while the rest of the AssetInfo interface deals in GUIDs. It also creates an ambiguity between the property `dependencies` and `GetDependencies`.

rhys-vdw commented:
Alright, addressed the above two points.

ogxd commented:
Hi @rhys-vdw, I look a little time to test your PRs and it seems to work nicely!    

A few remarks, if I may:
- PR is mixing different topics: using guids, adding prefixes to warnings and error management tweaks. Given the (non-existent) CI on this project, I'd say we can go for a yolo merge for convenience :) 
- It seems Unity now also exposes GUID as "real" guids (value types) as opposed to strings. I'm not sure if that would work with serialization but just to let you know I might play around with this a little before making the next release tag.

Thanks for your contribution!

rhys-vdw commented:
Hey. Thanks for that.

Yeah I saw the GUID type but I was unsure about serialization and also half way through the refactor.

The cache load error handling is there because this change caused an irrecoverable error when loading the old cache format.

I shouldn't have snuck the prefixes in haha. I still want to eliminate those warnings entirely.

Actually there are a few things I want to fix/improve beyond this. Hope you don't mind me making more issues and PRs.

ogxd commented:
Hey sure do problem feel free! But just so you know, having put my nose back into this project, I think I'll be making some big changes soon.

rhys-vdw commented:
@ogxd sounds good! I'll create issues first so there shouldnt' be any surprises.

I have an idea for how to update deps when any object is edited so it doesn't require a project save each time. I'll make an issue with my notes at some point.
